### DISCUSSION TITLE
Issue Creation of float array

### DISCUSSION THREAD

**Discussion by remahl** [dvidelabs/flatcc#307] [Q&A] [answered]
Hello everyone,

I have a FlatBuffers schema that includes an array of unions (referred to as "item"). Each item contains an identifier and a data section. The data section can either be a primitive type, such as uint32, or an array of float.

```
namespace investigation;

table item_uint32
{
    signal_id:uint;
    data:uint32;
}

table item_array_float
{
    signal_id:uint;
    data:[float];
}

union item
{
    item_uint32,
    item_array_float
}

table domain_data 
{
    signal_id:uint;
    data:[item];
}
```

The flatbuffer structure is composed successfully for the primitive type uint32. However, when attempting to compose the float vector, we encounter the error: 
"_create_offset_vector_direct: Assertion `types[i] == 0' failed."

The code to create a float array:

```
...
#define VEC_SIZE 20
float data[VEC_SIZE] = {1.2, 1.3,...};

flatbuffers_float_vec_ref_t itemFloatArrayVec =
    investigation_item_array_float_data_create(&enc->builder, data, VEC_SIZE);  

investigation_item_array_float_data_add(&enc->builder, itemFloatArrayVec);

investigation_item_array_float_ref_t floatData =
    investigation_item_array_float_create(&enc->builder, 3 /* magic id */, itemFloatArrayVec);

investigation_item_union_ref_t itemFloatArrayVecUnionType = investigation_item_as_item_array_float(floatData);
...
```

The `investigation_item_array_float_data_create` returns always 0.

**Comment by mikkelfj** [comment]
Try to enable assertions in builder.c and see if you can get more information out of it. See source code comments at top.

**Comment by mikkelfj** [comment]
investigation_item_array_float_create
vs
investigation_item_array_float_data_create

From memory:
one creates a vector of that type. 
the other creates a vector of that type and stores it in the data field of an already open table, as a convenience.

If you do not have an open table, the error is probably due to the current stack frame type not being a table type, and the error return code NULL triggers the error you see later on.

If it works with integers it might just be that you mixed up the two versions with and without data suffix during testing.

**Reply by remahl** [reply, marked as answer]
Thank you for the hint.
We did some additional investigation and replaced

```
investigation_item_array_float_data_create(...)
investigation_item_array_float_data_add(...)
```

with 

```
ref = flatcc_builder_create_vector(...)
investigation_item_array_float_create(ref);
```

this seems to work.

**Comment by mikkelfj** [comment]
Sounds good.
I forgot to mention:

The general format is

```
<namespace>_<typename>[_vector ][_<verb>]
<namespace>_<tablename>_<fieldname>[_<verb>]  (with _get being the default verb)
```

I might not remember exactly, but that is the idea. That is, all types also have a vector type regardless of whether a vector is used in the schema.

The variant with a field name knows whether it is a vector or not, and has different verbs, but is just a thin wrapper around the version without a field name. If they appear to have similar names that comes down to how the table is named.


Using `flatcc_builder_create_vector(...)` is valid, but it is an untyped low-level operation that is not as type-safe as the intended version though type safety is limited in all cases. I recommend you try to find the prober name for your type.

In your case the proper name is probaly something like

```
flatbuffers_float32_vector_create
```

All the basic types have defined vector operations.

Listing all available names may also be helpful. The following doc section might help, although it has been a while since it was last tested, you will need clang C compiler:

https://github.com/dvidelabs/flatcc?tab=readme-ov-file#extracting-documentation
